### PR TITLE
Opcua 1034 asynchronous methods

### PR DESCRIPTION
**Pull request by pnikiel**
Please review it when #46 and #47 are merged.

### CODE REVIEW

**Comment by pnikiel** [comment]
Ready for review, I think.

File: Design/Design.xsd [hunk 1]
```diff
@@ -395,25 +395,37 @@ elementFormDefault="qualified">
...
-    		<element name="returnvalue" type="tns:MethodReturnValue" minOccurs="0" maxOccurs="unbounded"></element>
+        <sequence>
+            <element name="argument" type="tns:MethodArgument" minOccurs="0" maxOccurs="unbounded"></element>
+            <element name="returnvalue" type="tns:MethodReturnValue" minOccurs="0" maxOccurs="unbounded"></element>
             <element name="documentation" type="tns:Documentation" maxOccurs="1" minOccurs="0"></element>
-    	</sequence>
-    	<attribute name="name" type="tns:VariableName" use="required"></attribute>
+        </sequence>
+        <attribute name="name" type="tns:VariableName" use="required"></attribute>
+        <attribute name="executionContext" type="tns:MethodExecutionContext" use="required"></attribute>
```
Comment (ben-farnham) [inline Design/Design.xsd]: Suggestion: how about calling this 'executionSynchronicity'? It might help to make it more apparent to the casual reader that this attribute concerns only the synchronous-ness of the method (my opinion: context is a rather vague term).

Comment (pnikiel) [reply Design/Design.xsd]: OK! And since you are the only native speaker among us, it makes double sense ;-)

Have you got any other thoughts? Or that's the only one?

Comment (ben-farnham) [reply Design/Design.xsd]: I didn't see anything else. I do find XSLT pretty hard to parse though. I'll try and make a quick test/demo of the functionality for myself tomorrow. I'll post again after that.

**Comment by pnikiel** [comment]
OK let me know when you have some results.

**Comment by pnikiel** [comment]
any news on this, @ben-farnham ?

**Comment by ben-farnham** [comment]
So, only tried with open6 so far but seems to fail on the two counts below:

image

Presumably there's no thread_pool or argsMissing status in open6-compat master?

**Comment by ben-farnham** [comment]
I'll add the missing status code to open6, but the thread pool stuff looks trickier...

**Comment by pnikiel** [comment]
Hi Ben,

I'll have a look ... I actually tested it with open6 so I'm not sure where it failed. Apparently you tested for a different case which, obviously, brings the highest value. Any chance you attach your Design?

**Comment by ben-farnham** [comment]
The design.xml

```
<?xml version="1.0" encoding="UTF-8"?>
<d:design projectShortName="" xmlns:d="http://cern.ch/quasar/Design"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://cern.ch/quasar/Design Design.xsd ">

	<d:class name="AsyncMethodsTestClass">
		<d:devicelogic/>
		<d:method name="asyncMethod" executionSynchronicity="asynchronous">
			<d:argument dataType="OpcUa_Int32" name="someUInt32Value" />
			<d:argument dataType="UaString" name="someStringValue" />
		</d:method>
	</d:class>

	<d:root>
		<d:hasobjects instantiateUsing="design" class="AsyncMethodsTestClass">
			<d:object name="AsyncMethodsTestClass"/>
		</d:hasobjects>
	</d:root>

</d:design>
```

**Comment by ben-farnham** [comment]
Added the status code mapping in open62541-compat. Ready for merge (single line addition) to master.
https://github.com/quasar-team/open62541-compat/pull/30

**Comment by pnikiel** [comment]
@ben-farnham I will have to ask you to investigate the -lpthread problem. I verified and it seems to be passed from the command line. Perhaps you use some custom build config?

**Comment by pnikiel** [comment]
@ben-farnham this seems to be stuck on the -lpthread problem reported by Travis. What can we do to move on?
